Unable to open asar archive when having no execution permission on $HOME

I'm having a strange issue in some GNU/Linux systems with v0.35.2:

```sh
Uncaught Error: Cannot find module '/home/jviotti/Downloads/7193eec573c62867151d-4e63199c2cf83a13e4a2301c61deec2616f9187b/node_modules/electron-prebuilt/dist/resources/atom.asar/renderer/lib/init.js'
```

![screenshot from 2015-12-02 13-35-42](https://cloud.githubusercontent.com/assets/2192773/11538957/ea73cbf6-98fa-11e5-9f27-7324bd6e39e6.png)

The file referenced above *does* exist:

```sh
$ asar list node_modules/electron-prebuilt/dist/resources/atom.asar | grep renderer/lib/init.js
/renderer/lib/init.js
```

I've been able to build a small dummy application that reproduces the error: https://gist.github.com/jviotti/7193eec573c62867151d.

#### Instructions:

1. Download the gist as a zip.
2. `npm install`.
3. Execute `npm start` as sudo: `sudo npm start` (the app runs fine without `sudo`).

#### Some experimentation I did:

- I can reproduce the issue on Fedora 22, but not on an Ubuntu 14.04 virtual machine.
- The issue is gone by downgrading to v0.31.2 (any higher version fails).
- The issue happens by using `npm start` as well as executing the `electron` binary directly.

I'm bisecting between v0.32.0 and v0.31.2 but `bootstrap.py` is taking forever in each step here (I don't have a good internet connection currently).

What could be causing this?

These are the errors I get on the broken commits mentioned above:

```
parallels@ubuntu:~/Projects/electron$ ./script/build.py -c D
ninja: Entering directory `out/D'
[18/1114] CXX obj/vendor/brightray/browser/brightray.browser_client.o
FAILED: /home/parallels/Projects/electron/vendor/llvm-build/Release+Asserts/bin/clang++ -MMD -MF obj/vendor/brightray/browser/brightray.browser_client.o.d -DNDEBUG -DV8_USE_EXTERNAL_STARTUP_DATA -DSK_SUPPORT_LEGACY_GETTOPDEVICE -DSK_SUPPORT_LEGACY_BITMAP_CONFIG -DSK_SUPPORT_LEGACY_DEVICE_VIRTUAL_ISOPAQUE -DSK_SUPPORT_LEGACY_N32_NAME -DSK_SUPPORT_LEGACY_SETCONFIG -DSK_IGNORE_ETC1_SUPPORT -DSK_IGNORE_GPU_DITHER -DTOOLKIT_VIEWS -DUSE_AURA -DUSE_X11 -DUSE_NSS_CERTS -DUSE_NSS -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUG -DCOMPONENT_BUILD -DGURL_DLL -DSKIA_DLL -DUSING_V8_SHARED -DWEBKIT_DLL -I../../vendor/brightray -I/home/parallels/Projects/electron/vendor/brightray/vendor/download/libchromiumcontent/src -I/home/parallels/Projects/electron/vendor/brightray/vendor/download/libchromiumcontent/src/skia/config -I/home/parallels/Projects/electron/vendor/brightray/vendor/download/libchromiumcontent/src/third_party/skia/include/core -I/home/parallels/Projects/electron/vendor/brightray/vendor/download/libchromiumcontent/src/third_party/mojo/src -I/home/parallels/Projects/electron/vendor/brightray/vendor/download/libchromiumcontent/src/third_party/WebKit -I/home/parallels/Projects/electron/vendor/brightray/vendor/download/libchromiumcontent/shared_library/gen -pthread -I/usr/include/gtk-2.0 -I/usr/lib/x86_64-linux-gnu/gtk-2.0/include -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/pango-1.0 -I/usr/include/gio-unix-2.0/ -I/usr/include/freetype2 -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/pixman-1 -I/usr/include/libpng12 -I/usr/include/harfbuzz -I/usr/include/dbus-1.0 -I/usr/lib/x86_64-linux-gnu/dbus-1.0/include -I/usr/include/gconf/2 -I/usr/include/nss -I/usr/include/nspr -Wno-deprecated-register -Wno-sentinel -m64 -march=x86-64 -std=c++11 -Wno-reserved-user-defined-literal -D__STRICT_ANSI__ -fno-rtti  -c ../../vendor/brightray/browser/browser_client.cc -o obj/vendor/brightray/browser/brightray.browser_client.o
In file included from ../../vendor/brightray/browser/browser_client.cc:7:
In file included from ../../vendor/brightray/browser/browser_context.h:8:
../../vendor/brightray/browser/permission_manager.h:26:72: error: non-virtual member function marked 'override' hides virtual member function
      const base::Callback<void(content::PermissionStatus)>& callback) override;
                                                                       ^
/home/parallels/Projects/electron/vendor/brightray/vendor/download/libchromiumcontent/src/content/public/browser/permission_manager.h:29:16: note: hidden overloaded virtual function 'content::PermissionManager::RequestPermission' declared here: type mismatch at 2nd parameter ('content::RenderFrameHost *' vs 'content::WebContents *')
  virtual void RequestPermission(
               ^
In file included from ../../vendor/brightray/browser/browser_client.cc:7:
In file included from ../../vendor/brightray/browser/browser_context.h:8:
../../vendor/brightray/browser/permission_manager.h:30:63: error: non-virtual member function marked 'override' hides virtual member function
                               const GURL& requesting_origin) override;
                                                              ^
/home/parallels/Projects/electron/vendor/brightray/vendor/download/libchromiumcontent/src/content/public/browser/permission_manager.h:39:16: note: hidden overloaded virtual function 'content::PermissionManager::CancelPermissionRequest' declared here: type mismatch at 2nd parameter ('content::RenderFrameHost *' vs 'content::WebContents *')
  virtual void CancelPermissionRequest(PermissionType permission,
               ^
2 errors generated.
[18/1114] CXX obj/vendor/brightray/browser/brightray.browser_context.o
FAILED: /home/parallels/Projects/electron/vendor/llvm-build/Release+Asserts/bin/clang++ -MMD -MF obj/vendor/brightray/browser/brightray.browser_context.o.d -DNDEBUG -DV8_USE_EXTERNAL_STARTUP_DATA -DSK_SUPPORT_LEGACY_GETTOPDEVICE -DSK_SUPPORT_LEGACY_BITMAP_CONFIG -DSK_SUPPORT_LEGACY_DEVICE_VIRTUAL_ISOPAQUE -DSK_SUPPORT_LEGACY_N32_NAME -DSK_SUPPORT_LEGACY_SETCONFIG -DSK_IGNORE_ETC1_SUPPORT -DSK_IGNORE_GPU_DITHER -DTOOLKIT_VIEWS -DUSE_AURA -DUSE_X11 -DUSE_NSS_CERTS -DUSE_NSS -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUG -DCOMPONENT_BUILD -DGURL_DLL -DSKIA_DLL -DUSING_V8_SHARED -DWEBKIT_DLL -I../../vendor/brightray -I/home/parallels/Projects/electron/vendor/brightray/vendor/download/libchromiumcontent/src -I/home/parallels/Projects/electron/vendor/brightray/vendor/download/libchromiumcontent/src/skia/config -I/home/parallels/Projects/electron/vendor/brightray/vendor/download/libchromiumcontent/src/third_party/skia/include/core -I/home/parallels/Projects/electron/vendor/brightray/vendor/download/libchromiumcontent/src/third_party/mojo/src -I/home/parallels/Projects/electron/vendor/brightray/vendor/download/libchromiumcontent/src/third_party/WebKit -I/home/parallels/Projects/electron/vendor/brightray/vendor/download/libchromiumcontent/shared_library/gen -pthread -I/usr/include/gtk-2.0 -I/usr/lib/x86_64-linux-gnu/gtk-2.0/include -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/pango-1.0 -I/usr/include/gio-unix-2.0/ -I/usr/include/freetype2 -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/pixman-1 -I/usr/include/libpng12 -I/usr/include/harfbuzz -I/usr/include/dbus-1.0 -I/usr/lib/x86_64-linux-gnu/dbus-1.0/include -I/usr/include/gconf/2 -I/usr/include/nss -I/usr/include/nspr -Wno-deprecated-register -Wno-sentinel -m64 -march=x86-64 -std=c++11 -Wno-reserved-user-defined-literal -D__STRICT_ANSI__ -fno-rtti  -c ../../vendor/brightray/browser/browser_context.cc -o obj/vendor/brightray/browser/brightray.browser_context.o
In file included from ../../vendor/brightray/browser/browser_context.cc:5:
In file included from ../../vendor/brightray/browser/browser_context.h:8:
../../vendor/brightray/browser/permission_manager.h:22:16: error: no type named 'WebContents' in namespace 'content'
      content::WebContents* web_contents,
      ~~~~~~~~~^
../../vendor/brightray/browser/permission_manager.h:28:41: error: no type named 'WebContents' in namespace 'content'
                               content::WebContents* web_contents,
                               ~~~~~~~~~^
In file included from ../../vendor/brightray/browser/browser_context.cc:8:
../../vendor/brightray/browser/inspectable_web_contents_impl.h:149:68: error: non-virtual member function marked 'override' hides virtual member function
      content::SessionStorageNamespace* session_storage_namespace) override;
                                                                   ^
/home/parallels/Projects/electron/vendor/brightray/vendor/download/libchromiumcontent/src/content/public/browser/web_contents_delegate.h:293:16: note: hidden overloaded virtual function 'content::WebContentsDelegate::ShouldCreateWebContents' declared here: type mismatch at 5th parameter ('const std::string &' (aka 'const basic_string<char> &') vs 'const base::string16 &' (aka 'const basic_string<char16, base::string16_char_traits> &'))
  virtual bool ShouldCreateWebContents(
               ^
In file included from ../../vendor/brightray/browser/browser_context.cc:9:
../../vendor/brightray/browser/network_delegate.h:64:67: error: only virtual member functions can be marked 'override'
  bool OnCanThrottleRequest(const net::URLRequest& request) const override;
                                                                  ^~~~~~~~
../../vendor/brightray/browser/browser_context.cc:188:35: error: allocating an object of abstract class type 'brightray::PermissionManager'
    permission_manager_.reset(new PermissionManager);
                                  ^
/home/parallels/Projects/electron/vendor/brightray/vendor/download/libchromiumcontent/src/content/public/browser/permission_manager.h:29:16: note: unimplemented pure virtual method 'RequestPermission' in 'PermissionManager'
  virtual void RequestPermission(
               ^
/home/parallels/Projects/electron/vendor/brightray/vendor/download/libchromiumcontent/src/content/public/browser/permission_manager.h:39:16: note: unimplemented pure virtual method 'CancelPermissionRequest' in 'PermissionManager'
  virtual void CancelPermissionRequest(PermissionType permission,
               ^
5 errors generated.
[18/1114] CXX obj/vendor/brightray/browser/brightray.browser_main_parts.o
FAILED: /home/parallels/Projects/electron/vendor/llvm-build/Release+Asserts/bin/clang++ -MMD -MF obj/vendor/brightray/browser/brightray.browser_main_parts.o.d -DNDEBUG -DV8_USE_EXTERNAL_STARTUP_DATA -DSK_SUPPORT_LEGACY_GETTOPDEVICE -DSK_SUPPORT_LEGACY_BITMAP_CONFIG -DSK_SUPPORT_LEGACY_DEVICE_VIRTUAL_ISOPAQUE -DSK_SUPPORT_LEGACY_N32_NAME -DSK_SUPPORT_LEGACY_SETCONFIG -DSK_IGNORE_ETC1_SUPPORT -DSK_IGNORE_GPU_DITHER -DTOOLKIT_VIEWS -DUSE_AURA -DUSE_X11 -DUSE_NSS_CERTS -DUSE_NSS -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUG -DCOMPONENT_BUILD -DGURL_DLL -DSKIA_DLL -DUSING_V8_SHARED -DWEBKIT_DLL -I../../vendor/brightray -I/home/parallels/Projects/electron/vendor/brightray/vendor/download/libchromiumcontent/src -I/home/parallels/Projects/electron/vendor/brightray/vendor/download/libchromiumcontent/src/skia/config -I/home/parallels/Projects/electron/vendor/brightray/vendor/download/libchromiumcontent/src/third_party/skia/include/core -I/home/parallels/Projects/electron/vendor/brightray/vendor/download/libchromiumcontent/src/third_party/mojo/src -I/home/parallels/Projects/electron/vendor/brightray/vendor/download/libchromiumcontent/src/third_party/WebKit -I/home/parallels/Projects/electron/vendor/brightray/vendor/download/libchromiumcontent/shared_library/gen -pthread -I/usr/include/gtk-2.0 -I/usr/lib/x86_64-linux-gnu/gtk-2.0/include -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/pango-1.0 -I/usr/include/gio-unix-2.0/ -I/usr/include/freetype2 -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/pixman-1 -I/usr/include/libpng12 -I/usr/include/harfbuzz -I/usr/include/dbus-1.0 -I/usr/lib/x86_64-linux-gnu/dbus-1.0/include -I/usr/include/gconf/2 -I/usr/include/nss -I/usr/include/nspr -Wno-deprecated-register -Wno-sentinel -m64 -march=x86-64 -std=c++11 -Wno-reserved-user-defined-literal -D__STRICT_ANSI__ -fno-rtti  -c ../../vendor/brightray/browser/browser_main_parts.cc -o obj/vendor/brightray/browser/brightray.browser_main_parts.o
In file included from ../../vendor/brightray/browser/browser_main_parts.cc:7:
In file included from ../../vendor/brightray/browser/browser_context.h:8:
../../vendor/brightray/browser/permission_manager.h:22:16: error: no type named 'WebContents' in namespace 'content'
      content::WebContents* web_contents,
      ~~~~~~~~~^
../../vendor/brightray/browser/permission_manager.h:28:41: error: no type named 'WebContents' in namespace 'content'
                               content::WebContents* web_contents,
                               ~~~~~~~~~^
2 errors generated.
ninja: build stopped: subcommand failed.
```

I've modified `Archive::Init` to print the mode of `path_.value()` before checking for it's validity:

```cpp
#include <sys/stat.h>

...

int getPathMode(std::string name) {
  struct stat buffer;   
  int ret = stat(name.c_str(), &buffer);
  if (ret < 0) return -1;
  return buffer.st_mode;
}

bool Archive::Init() {
  std::cout << getPathMode(path_.value()) << std::endl;
  if (!file_.IsValid()) {
    std::cout << "Archive is invalid: ("
              << file_.ErrorToString(file_.error_details())
              << "): "
              << path_.value()
              << std::endl;
    return false;
  }

...
```

This is the output when running my app as `root` (notice I packed the gist I linked above as `resources/app.asar` for convenience):

```sh
$ sudo ./out/D/electron 
33204
33204
(electron) loadUrl is deprecated. Use loadURL instead.
33204
-1
Archive is invalid: (FILE_ERROR_ACCESS_DENIED): /home/jviotti/electron-current/out/D/resources/atom.asar
-1
Archive is invalid: (FILE_ERROR_ACCESS_DENIED): /home/jviotti/electron-current/out/D/resources/atom.asar
-1
Archive is invalid: (FILE_ERROR_ACCESS_DENIED): /home/jviotti/electron-current/out/D/resources/atom.asar
-1
Archive is invalid: (FILE_ERROR_ACCESS_DENIED): /home/jviotti/electron-current/out/D/resources/atom.asar
-1
Archive is invalid: (FILE_ERROR_ACCESS_DENIED): /home/jviotti/electron-current/out/D/resources/atom.asar
-1
Archive is invalid: (FILE_ERROR_ACCESS_DENIED): /home/jviotti/electron-current/out/D/resources/atom.asar
-1
Archive is invalid: (FILE_ERROR_ACCESS_DENIED): /home/jviotti/electron-current/out/D/resources/atom.asar
-1
Archive is invalid: (FILE_ERROR_ACCESS_DENIED): /home/jviotti/electron-current/out/D/resources/atom.asar
-1
Archive is invalid: (FILE_ERROR_ACCESS_DENIED): /home/jviotti/electron-current/out/D/resources/atom.asar
-1
Archive is invalid: (FILE_ERROR_ACCESS_DENIED): /home/jviotti/electron-current/out/D/resources/atom.asar
-1
Archive is invalid: (FILE_ERROR_ACCESS_DENIED): /home/jviotti/electron-current/out/D/resources/atom.asar
-1
Archive is invalid: (FILE_ERROR_ACCESS_DENIED): /home/jviotti/electron-current/out/D/resources/atom.asar
-1
Archive is invalid: (FILE_ERROR_ACCESS_DENIED): /home/jviotti/electron-current/out/D/resources/atom.asar
-1
Archive is invalid: (FILE_ERROR_ACCESS_DENIED): /home/jviotti/electron-current/out/D/resources/atom.asar
-1
Archive is invalid: (FILE_ERROR_ACCESS_DENIED): /home/jviotti/electron-current/out/D/resources/atom.asar
-1
Archive is invalid: (FILE_ERROR_ACCESS_DENIED): /home/jviotti/electron-current/out/D/resources/atom.asar
[30652:0128/113606:ERROR:CONSOLE(340)] "Uncaught Error: Cannot find module '/home/jviotti/electron-current/out/D/resources/atom.asar/renderer/lib/init.js'", source: module.js (340)
```

Notice that the stat succeeds three times at the beginning, apparently before the `index.html` file gets loaded.

--- a/script/lib/config.py
+++ b/script/lib/config.py
@@ -8,7 +8,7 @@ import sys
 
 BASE_URL = os.getenv('LIBCHROMIUMCONTENT_MIRROR') or \
     'https://s3.amazonaws.com/github-janky-artifacts/libchromiumcontent'
-LIBCHROMIUMCONTENT_COMMIT = '1be1a55abba527d6fb0fc2a589cbaef2f5a6a79e'
+LIBCHROMIUMCONTENT_COMMIT = 'fc5174471ccccce7030d0960f5228e73ca7c2ac6'
 
 PLATFORM = {
   'cygwin': 'win32',

